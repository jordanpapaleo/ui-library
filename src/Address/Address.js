import React, {Component, Children} from 'react'
import PropTypes from 'prop-types'
const NO_DATA = '--'

function titleCase (str = NO_DATA) {
  str = str.replace(/\s+/g, ' ')
  str = str.replace(/\s+$/, '')

  return str
    .toLowerCase()
    .split(' ')
    .map(word => word.replace(word[0], word[0].toUpperCase()))
    .join(' ')
}

export default class Address extends Component {
  static propTypes = {
    address: PropTypes.shape({
      city: PropTypes.string,
      state: PropTypes.string,
      street: PropTypes.string,
      unitNumber: PropTypes.string,
      zipcode: PropTypes.string
    }).isRequired,
    children: PropTypes.any,
    twoLines: PropTypes.bool
  }

  static Street = ({address, twoLines, style}) => {
    const {street, unit} = address
    return <span style={style}>{street}{unit}{twoLines ? <br /> : ', '}</span>
  }

  static CityStateZip = ({address, style}) => {
    const {city, state, zip} = address
    return <span style={style}>{city}, {state} {zip}</span>
  }

  render () {
    const {address, twoLines, children} = this.props
    const street = titleCase(address.street)
    const unit = address.unitNumber ? ` ${titleCase(address.unitNumber)}` : ''
    const city = titleCase(address.city)
    const state = (address.state && address.state.length) ? address.state.toUpperCase() : NO_DATA
    const zip = (address.zipcode && address.zipcode.length) ? address.zipcode.substr(0, 5) : NO_DATA

    return children
      ? (
        <span className='address-component'>
          {Children.map(children, (child) => React.cloneElement(child, {
            style: child.props.style,
            address: {street, unit, city, state, zip},
            twoLines
          }))}
        </span>
      )
      : (
        <span className='address-component'>
          <Address.Street address={{street, unit, city, state, zip}} twoLines={twoLines} />
          <Address.CityStateZip address={{street, unit, city, state, zip}} />
        </span>
      )
  }
}
