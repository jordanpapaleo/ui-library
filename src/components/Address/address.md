# Address

## Params

```
address: PropTypes.shape({
  city: PropTypes.string,
  state: PropTypes.string,
  street: PropTypes.string,
  unitNumber: PropTypes.string,
  zipcode: PropTypes.string
}).isRequired

children: PropTypes.any

twoLines: PropTypes.bool
```

## Code Example

```javascript
import {Address} from '@jordanpapaleo/ui-library'
```

### Single line

```javascript
<Address address={address} />
```

### Multi Line

```javascript
<Address address={address} twoLines />
```

### Compound Component

```javascript
<Address address={address} twoLines>
  <Address.Street style={{fontWeight: 'bold', fontSize: 18, lineHeight: 1.1}} />
  <Address.CityStateZip style={{color: 'orange', fontSize: 16, lineHeight: 1.1}} />
</Address>
```
