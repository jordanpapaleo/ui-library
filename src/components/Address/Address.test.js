import React from "react";
import ReactDOM from "react-dom";
import Address from "./Address";
import renderer from "react-test-renderer";

const address = {
  street: "123 Sesame Street",
  city: "New York",
  state: "NY",
  zipcode: "12345-6789"
};

describe("Address", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");

    ReactDOM.render(<Address address={address} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("matches snapshot", () => {
    const tree = renderer.create(<Address address={address} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should test Street function", () => {
    const wrapper = new Address({ address });
    console.log(wrapper);
    const result = Address.Street({ address, twoLines: true });
    console.log(result);
  });
});
