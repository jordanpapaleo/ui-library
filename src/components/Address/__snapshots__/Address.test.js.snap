// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Address matches snapshot 1`] = `
<span
  className="address-component"
>
  <span
    style={undefined}
  >
    123 Sesame Street
    
    , 
  </span>
  <span
    style={undefined}
  >
    New York
    , 
    NY
     
    12345
  </span>
</span>
`;
